internal/controller/rabbitmq/rabbitmquser_controller.go

```diff
@@ -0,0 +1,401 @@
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	h, _ := helper.NewHelper(instance, r.Client, r.Kclient, r.Scheme, Log)
+
+	// Save a copy of the conditions so that we can restore the LastTransitionTime
+	// when a condition's state doesn't change
+	savedConditions := instance.Status.Conditions.DeepCopy()
+
+	// Initialize status conditions
+	cl := condition.CreateList(
+		condition.UnknownCondition(condition.ReadyCondition, condition.InitReason, condition.ReadyInitMessage),
+		condition.UnknownCondition(rabbitmqv1.RabbitMQUserReadyCondition, condition.InitReason, rabbitmqv1.RabbitMQUserReadyInitMessage),
+	)
+	instance.Status.Conditions.Init(&cl)
+	instance.Status.ObservedGeneration = instance.Generation
+
+	defer func() {
+		// Restore condition timestamps if they haven't changed
+		condition.RestoreLastTransitionTimes(&instance.Status.Conditions, savedConditions)
+
+		if instance.Status.Conditions.IsUnknown(condition.ReadyCondition) {
+			instance.Status.Conditions.Set(instance.Status.Conditions.Mirror(condition.ReadyCondition))
+		}
+		if err := h.PatchInstance(ctx, instance); err != nil {
+			Log.Error(err, "Failed to patch instance")
+		}
+	}()
+
+	// Handle deletion
+	if !instance.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, instance, h)
+	}
+
+	// Add finalizer if not being deleted
+	if controllerutil.AddFinalizer(instance, userFinalizer) {
+		// Finalizer was added, update will trigger reconcile
+		return ctrl.Result{}, nil
+	}
+
+	return r.reconcileNormal(ctx, instance, h)
+}
+
+func (r *RabbitMQUserReconciler) reconcileNormal(ctx context.Context, instance *rabbitmqv1.RabbitMQUser, h *helper.Helper) (ctrl.Result, error) {
+	Log := log.FromContext(ctx)
+
+	// Username is defaulted by webhook
+	username := instance.Spec.Username
+
+	// Handle VhostRef changes - remove finalizer from old vhost if changed
+	// We track the previous vhost CR name in status.VhostRef to detect changes
+	userFinalizer := rabbitmqv1.UserVhostFinalizerPrefix + instance.Name
+	if instance.Status.VhostRef != "" && instance.Status.VhostRef != instance.Spec.VhostRef {
+		// VhostRef changed - remove finalizer from old vhost
+		oldVhost := &rabbitmqv1.RabbitMQVhost{}
+		if err := r.Get(ctx, types.NamespacedName{Name: instance.Status.VhostRef, Namespace: instance.Namespace}, oldVhost); err == nil {
+			if controllerutil.RemoveFinalizer(oldVhost, userFinalizer) {
+				if err := r.Update(ctx, oldVhost); err != nil {
+					// Requeue to retry - this is important for VhostRef changes
+					Log.Error(err, "Failed to remove finalizer from old vhost, requeueing", "vhost", instance.Status.VhostRef, "finalizer", userFinalizer)
+					return ctrl.Result{RequeueAfter: time.Duration(2) * time.Second}, err
+				}
+			}
+		} else if !k8s_errors.IsNotFound(err) {
+			Log.Error(err, "Failed to get old vhost for finalizer removal", "vhost", instance.Status.VhostRef)
+			// Don't fail here - old vhost might have been deleted
```

Comment:
not sure if we already discussed this. do we have to return here if we can not get the oldVhost and it is not IsNotFound? if the old vhost might been deleted, we should get an IsNotFound?